ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java

```diff
@@ -327,12 +332,30 @@ private boolean shouldTakeSnapshot() {
     if (autoSnapshotThreshold == null) {
       return false;
     } else if (shouldStop()) {
-      return triggerSnapshotWhenStopEnabled && getLastAppliedIndex() - snapshotIndex.get() > 0;
+      return shouldTakeSnapshotAccordingToConfAndStatus() && getLastAppliedIndex() - snapshotIndex.get() > 0;
     }
     return state == State.RUNNING &&
         getStateMachineLastAppliedIndex() - snapshotIndex.get() >= autoSnapshotThreshold;
   }
 
+  /**
+   * In view of the three variables triggerSnapshotWhenStopEnabled, triggerSnapshotWhenRemoveEnabled and isRemoving,
+   * we can draw the following 8 combination:
+   * true true true => true
+   * true true false => true
+   * true false true => false
+   * true false false => true
+   * false true true => true
+   * false true false =>  false
+   * false false true => false
+   * false false false => false
+   * @return result
+   */
+  private boolean shouldTakeSnapshotAccordingToConfAndStatus() {
+    return (triggerSnapshotWhenStopEnabled || isRemoving)
+        && (triggerSnapshotWhenRemoveEnabled || !isRemoving);
```

Comment:
Let's call it `shouldTakeSnapshotAtStop`.  We may simply the logic as below:
```java
  private boolean shouldTakeSnapshotAtStop() {
    if (isRemoving) {
      return triggerSnapshotWhenRemoveEnabled;
    } else {
      return triggerSnapshotWhenStopEnabled;
    }
  }
```

Reply:
The codes looks much clearer!